First half of fix for ticket 6260.

odersky opened the pull request:
This commit fixes the first half of SI-6260. Duplicate methods are no longer generated.
We get another problem (a ClassCast exeption) which is a much deeper problem. 

It looks like we will have to disallow again value classes that wrap over a type parameter. Folliowing up in scala-sips.

Review by @paulp @retronym

odersky committed:
Fixes SI-6260

Guards against bridge methods that clash with other methods. Two tests: The neg test is the original ticket. The run test tweaks things slightly so that the generated bridge method does not clash, and tests that the necessary unboxings are indeed performed at runtime.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/405/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1112/

retronym commented:
I'm not deeply enough into these phases to judge correctness. But this looks very fiddly to get right and keep right, and might benefit from comments near the places that are time travelling back and forth between nearby phases.

``` scala
// We have to look after post erasure for ... so that we can see the effect of xxx
// But, ... must be checked before these transformations
import self.{afterPostErasure => xxx, afterErasure => yyy}

xxx {

}
```

The alias only really makes sense if you can find a name to describe the local intent of that phase lookup, which is used in a few places below. But a comment should be included in any case.

You had to change a few calls to `erasure` to `postErasure`. Do you believe you've got them all, or should we review the rest of the usages? (I've only access to the web UI right now, so I haven't gone digging myself.)

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/405/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1112/

gkossakowski commented:
Also, what about test coverage? Especially because the diff looks tricky.

odersky committed:
Cleanup of OverridingPairs

Added a default value to a hashmap so that we can work with it without Option acrobatics.
